### PR TITLE
Add an integer date format to a support enumeration and comparison

### PR DESCRIPTION
**Pull request by kipcole9**
From a conversation on Elixir Forum came a discussion about calendar interoperability and conversion.  This PR is in response to that conversation.  

Fundamentally is adds the idea of an "integer date" which can be used for the basis of conversion between calendars.  For dates in the default `Calendar.ISO` there is no conversion hence no overhead.  

This commit adds:

* An integer date format and to/from conversions as part of the `Calendar` behaviour that
supports the means to convert dates between calendars.  The mechanism is derived from
Calendrical Calculations written by Dershowitz and Rheingold.

* A set of `kday` calculations that allow calendar independent calculation of `first`,
`last`, `nth`, `on`, `before`, `after` days

* `Date.diff/2` to calculate the difference in days between two dates from any calendar

* Update to the Calendar behaviour to define the requirement for a `to_integer_date/1`
and `from_integer_date/1` that underpins calendar calculations and conversions

* Add support for the Enum protocol for a date range to allow enumeration over a
range of dates (possible because of the internal conversion to integer dates)

* A Julian Calendar as an example of an a calendar implementation using the revised
behaviour and calendar functions

* Tests for the above

### CODE REVIEW

**Comment by josevalim** [comment]
Thank you @kipcole9, this looks excellent. As there are too many features here, we may end-up merging them in part, also to facilitate discussion. In the first round, I will focus only on the Calendar callbacks, and then we can move up.

I will likely not merge the Julian calendar although we can move it to tests in order to check the functionality. Don't worry about doing any of those changes for now though. :)

File: lib/elixir/lib/calendar.ex [hunk 1]
```diff
@@ -194,6 +234,16 @@ defmodule Date do
   end
 
   @doc """
+  Returns the number of days in a week.
+
+  Always returns 7 days in a week irrespective of the
+  calendar in use.
+  """
+  def days_in_a_week() do
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: Should we impose this constraint? Which calendars will we reject by making such choice? Why not ask the calendar for this value?

File: lib/elixir/lib/calendar.ex [hunk 2]
```diff
@@ -417,12 +489,273 @@ defmodule Date do
...
       2
+
   """
-  @spec day_of_week(Calendar.date) :: non_neg_integer()
+  @spec day_of_week(Calendar.date | Calendar.integer_date) :: non_neg_integer()
   def day_of_week(%{calendar: calendar, year: year, month: month, day: day}) do
     calendar.day_of_week(year, month, day)
   end
 
+  def day_of_week(date) when is_integer(date) do
```
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: I am not sure if we should expose functions such as this. I would rather have developers make the conversion explicit maybe by introducing something such as `Date.from_days(...)` or something similar.

**Comment by josevalim** [comment]
@kipcole9 I have added some questions in this first review pass. I believe someone proposed to use the Unix Epoch but I believe picking one or the other would not cause any relevant difference in the code, would it? If so, I would prefer to keep the current epoch. What are your thoughts?

**Comment by josevalim** [comment]
I have removed the earlier question since it is already being discussed in this thread: https://elixirforum.com/t/calendars-calendar-conversions/3119/2

File: lib/elixir/lib/calendar.ex (hunk 2 above)
Comment (kipcole9) [inline lib/elixir/lib/calendar.ex]: José, given the discussion in ElixirForum I'm not sure the right strategy.  In part it depends on whether the intermediate format of `{RD/JD, milliseconds}` is a first class citizen or not.  I don't have a super strong view on that - but it seems we may actually end up with a date format that is convertible and monotonic which would seem to be a good thing.

For now I'm going to update my code to use the new proposed intermediate format and resubmit and we can see where is goes from there - unless there is already a strong view?

File: lib/elixir/lib/calendar.ex (hunk 1 above)
Comment (kipcole9) [inline lib/elixir/lib/calendar.ex]: José, this is a bit lazy of me on two fronts:

1.  All of the calendars in D&R that use a format `{year, month, day}` are 7 days in a week. 
2. For calculations on `{RD, milliseconds}` (or my original case of `integer`} like the `kday` functions, there is no longer the concept of a calendar so an assumption has to be made.

I suspect you may consider the `kday` functions too much for the expressed intent of the core Calendar module and therefore I can certainly extract them.  But they, and the Enum protocol support, are extremely useful for anyone doing calendar work (calculating public holidays / festivals / religious and cultural events) which I why I proposed including them.

Comment (kipcole9) [inline lib/elixir/lib/calendar.ex]: Also did you mean to suggest `{RD, milliseconds}` or `{RD, microseconds}`?.  `Elixir.Time` is microseconds resolution so I assume that would be the intent for consistency?

File: lib/elixir/lib/calendar.ex (hunk 2 above)
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: > For now I'm going to update my code to use the new proposed intermediate format and resubmit and we can see where is goes from there - unless there is already a strong view?

That sounds a great idea but please don't "force push". Just add commits on top so we can revert the changes in the chance we find out that using `{RD/JD, fraction}` is a bad idea.

File: lib/elixir/lib/calendar.ex (hunk 1 above)
Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: I meant to suggest `{RD, microseconds}`. We are having a discussion on the forum about this though.

> For calculations on {RD, microseconds} (or my original case of integer) like the kday functions, there is no longer the concept of a calendar so an assumption has to be made.

Sorry, I am confused. All of those functions receives the calendar. Couldn't we just ask the calendar how many days there are in a week before doing the conversion to/from integer?

It seems that the most important computation being done in the `kday` functions is to find how many days until the beginning of next week. So what if we introduce those computations and still make them calendar sensitive? For example:

    Date.days_until_beginning_of_week(date) + 2 - 1
What if we generalize the `kday` functions to `beginning_of_the_week` and `beginning_of_next_week`. Those will be calendar dependent function that receives a date and return a date that is the beginning of the same week (or next week). This way, we don't need to perform the rounding calculations and the `days_in_week` function should no longer be necessary. Developers can still achieve the "next monday" feature

Comment (josevalim) [inline lib/elixir/lib/calendar.ex]: In case you are reading my replies from e-mail, please ignore the lower half of my previous reply. It was there by accident. :)

Comment (Qqwy) [inline lib/elixir/lib/calendar.ex]: I agree that clarification is needed here. As for information on calendars that have other 'weeks' than 7-day-long ones, see [the Wikipedia article on weeks in other calendars.](https://en.wikipedia.org/wiki/Week#.22Weeks.22_in_other_calendars}. Most of them are not commonly used, but I believe we should consider them.

One of the calendars that uses a 10-day week is the French Revolutionary calendar, which I thought the D&R book covers? (I have not read the book (hopefully I can find it in the university library), but I have read their earlier two papers and their website with secondary content).

**Comment by lau** [comment]
Why not make a separate library first before adding anything to the standard library? The vast majority of people use the Gregorian calendar. If someone really need to use another Calendar everyday, they can make a separate library.

Even though it is interesting, we have to weigh it up against not complicating the standard library. In addition to that, once added it will be hard to change or remove.

**Comment by josevalim** [comment]
@lau I believe it is clear this is not guaranteed to be added to Elixir. Therefore, the PR is useful for us to see how it will impact the code and take an informed decision. If the conversion to an integer date ends-up being a conversion to gregorian, which we already do internally, then this is rather simple and a welcome addition which will allow us to implement other operations in a more straight-forward fashion. However, if we go towards a more complicated road, then it is probably best to let it develop in a separate project. We have already moved some initiatives, like supporting calendars that do not have yy-mm-dddd schemas, to separate libraries.

**Comment by josevalim** [comment]
Ping. :)

**Comment by Qqwy** [comment]
I'd still like to distill the constructive discussion we've had in the [Calendar Conversions topic](https://elixirforum.com/t/calendars-calendar-conversions/3119) in an updated or new Pull Request.

I sent @kipcole9 on the 16th of January and later on the 4th of February a personal message on the Elixir forum which he has not responded to yet.
If he wants to work further on this Pull Request (or on a new one), that would be splendid.
But if he is unable to for some reason, I'd be honored to work on it myself as well.
